tooling/greybox_fuzzer/src/corpus.rs

```diff
@@ -0,0 +1,431 @@
+    /// Creates the corpus directory if it doesn't exist.
+    /// Returns an error if directory creation fails or if any files can't be parsed.
+    pub fn load_corpus_from_disk(&mut self) -> Result<(), String> {
+        let mut builder = DirBuilder::new();
+        match builder.recursive(true).create(&self.corpus_path) {
+            Ok(_) => (),
+            Err(_) => {
+                return Err(format!(
+                    "Couldn't create directory {:?}",
+                    self.corpus_path.as_os_str()
+                ));
+            }
+        }
+        // Go through all files in the corpus directory recursively
+        for entry in WalkDir::new(&self.corpus_path) {
+            let Ok(entry) = entry else {
+                continue;
+            };
+            if !entry.file_type().is_file() {
+                continue;
+            }
+            if !entry.path().extension().map_or(false, |ext| ext == CORPUS_FILE_EXTENSION) {
+                continue;
+            };
+            let path = entry.into_path();
+            // Skip files that are already loaded
+            if self.file_manager.has_file(&path) {
+                continue;
+            }
+            let source = std::fs::read_to_string(path.as_path())
+                .unwrap_or_else(|_| panic!("could not read file {:?} into string", path));
+
+            let parsed_source = parse_json(&source, &self.abi).map_err(|parsing_error| {
+                format!("Error while parsing file {:?}: {:?}", path.as_os_str(), parsing_error)
+            })?;
+
+            // Add the file and its parsed contents to our tracking maps
+            let file_id = self.file_manager.add_file_with_source(path.as_path(), source).unwrap();
+            self.parsed_map.insert(file_id, parsed_source);
+        }
+        Ok(())
+    }
+
+    /// Returns a vector containing all parsed inputs from the corpus
+    pub fn get_full_corpus(&self) -> Vec<&InputMap> {
+        let file_ids = self.file_manager.as_file_map().all_file_ids();
+        let mut full_corpus = Vec::new();
+        for file_id in file_ids {
+            full_corpus.push(&self.parsed_map[file_id]);
+        }
+        full_corpus
+    }
+
+    /// Returns the path to the corpus directory
+    pub fn get_corpus_path(&self) -> &Path {
+        &self.corpus_path
+    }
+
+    /// Saves a new testcase to disk and adds it to the file manager.
+    /// The filename is generated from the SHA-256 hash of the contents.
+    /// Returns an error if file creation or writing fails.
+    pub fn save_testcase_to_disk(&mut self, contents: &str) -> Result<(), String> {
+        let mut builder = DirBuilder::new();
+        match builder.recursive(true).create(&self.corpus_path) {
+            Ok(_) => (),
+            Err(_) => {
+                return Err(format!(
+                    "Couldn't create directory {:?}",
+                    self.corpus_path.as_os_str()
+                ));
+            }
+        }
```

Comment:
This could be moved into a helper method (it also exists in `load_corpus_from_disk`)

Reply:
Maybe in a following PR